crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,529 @@
+                            .map(|(token, price)| NativePriceResult {
+                                token,
+                                result: price,
+                            })
+                            .collect(),
+                        Err(err) => {
+                            tracing::error!(?err, "failed to send native price batch request");
+                            batch
+                                .into_iter()
+                                .map(|token| NativePriceResult {
+                                    token,
+                                    result: Err(err.clone()),
+                                })
+                                .collect()
+                        }
+                    };
+                    for result in results {
+                        let _ = broadcast_sender.send(result);
+                    }
+                }
+            }
+        }))
+    }
+
+    /// Request to get estimate prices in a batch
+    pub async fn request_buffered_estimate_prices(
+        &self,
+        token: &H160,
+    ) -> NativePriceEstimateResult {
+        // Sends the token for requesting price
+        self.calls.unbounded_send(*token).map_err(|e| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "failed to append a new token to the queue: {e:?}"
+            ))
+        })?;
+
+        let mut rx = self.broadcast_sender.subscribe();
+
+        tokio::time::timeout(self.config.result_ready_timeout, async {
+            loop {
+                if let Ok(Some(result)) =
+                    Self::receive_with_timeout(&mut rx, token, self.config.batch_delay).await
+                {
+                    return result.result;
+                }
+            }
+        })
+        .await
+        .map_err(|_| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "blocking buffered estimate prices timeout elapsed"
+            ))
+        })?
+    }
+
+    // Function to receive with a timeout
+    async fn receive_with_timeout(
+        rx: &mut broadcast::Receiver<NativePriceResult>,
+        token: &H160,
+        timeout_duration: Duration,
+    ) -> Result<Option<NativePriceResult>, Elapsed> {
+        tokio::time::timeout(timeout_duration, async {
+            match rx.recv().await {
+                Ok(value) => (value.token == *token).then_some(value),
+                Err(_) => None,
+            }
+        })
+        .await
+    }
+}
+
+/// Batches a stream into chunks.
+///
+/// This is very similar to `futures::stream::StreamExt::ready_chunks` with the
+/// difference that it allows configuring a minimum delay for a batch, so
+/// waiting for a small amount of time to allow the stream to produce additional
+/// items, thus decreasing the chance of batches of size 1.
+fn batched_for_each<T, St, F, Fut>(
+    config: Configuration,
+    items: St,
+    work: F,
+) -> impl Future<Output = ()>
```

Comment:
Not related to this PR. Some of the code is copied from the `ethrpc` crate. Any plan to generalize it even further to avoid code duplications?

Reply:
@squadgazzz this is a good question. AFAIK, I was told we don't want to keep adding things to the `shared` crate, and the code is in different crates, so it makes it hard to share. We could have a super generic `buffered` crate, and then use it for both implementations.